scan-payment/src/main/java/com/getbouncer/scan/payment/analyzer/NameDetectAnalyzer.kt

```diff
@@ -0,0 +1,268 @@
+    )
+
+    override val name: String = "card_side_analyzer"
+
+    override suspend fun analyze(data: SSDOcr.SSDOcrInput, state: PaymentCardOcrState): Output {
+        return if (state.runNameExtraction && ssdObjectDetect != null && alphabetDetect != null) {
+            val objDetectInput = SSDObjectDetect.Input(
+                fullImage = data.fullImage,
+                previewSize = data.previewSize,
+                cardFinder = data.cardFinder,
+                iin = null
+            )
+
+            val output = ssdObjectDetect.analyze(objDetectInput, Unit)
+            var nameRect: RectF? = null
+
+            // check to see if we detected a legal name. If so, run our name extraction model, serially for now
+            for (box in output.detectionBoxes) {
+                if (box.label == SSDObjectDetect.Labels.NAME.ordinal) {
+                    nameRect = box.rect
+                    break
+                }
+            }
+
+            val name = if (nameRect != null) {
+                processPredictions(nameRect, objDetectInput.toObjectDetectionCroppedImage())
+            } else {
+                null
+            }
+
+            Output(name, output.detectionBoxes)
+        } else {
+            Output("", null)
+        }
+    }
+
+    internal class CharPredictionWithBox(
+        val characterPrediction: AlphabetDetect.Prediction,
+        val box: RectF
+    ) {
+        private val boxArray: FloatArray = FloatArray(4)
+
+        fun getPercentBoxArray(width: Int, height: Int): FloatArray {
+            val pctBoxArray = FloatArray(4)
+            pctBoxArray[0] = boxArray[0] / width
+            pctBoxArray[1] = boxArray[1] / height
+            pctBoxArray[2] = boxArray[2] / width
+            pctBoxArray[3] = boxArray[3] / height
+            return pctBoxArray
+        }
+
+        init {
+            boxArray[0] = box.left
+            boxArray[1] = box.top
+            boxArray[2] = box.right
+            boxArray[3] = box.bottom
+        }
+    }
+
+    private suspend fun processPredictions(
+        nameRect: RectF,
+        bitmapForObjectDetection: Bitmap
+    ): String? {
+        if (alphabetDetect == null) {
+            return null;
+        }
+        var x = (nameRect.left * bitmapForObjectDetection.width).toInt()
+        val y = (nameRect.top * bitmapForObjectDetection.height).toInt()
+        var width = (nameRect.width() * bitmapForObjectDetection.width).toInt()
+        val height = (nameRect.height() * bitmapForObjectDetection.height).toInt()
+
+        // adjust start and end
+        x -= height / 2
+        x = Math.max(0, x)
+        width += height
+        width = Math.min(bitmapForObjectDetection.width - x, width)
+
+        val nameBitmap = Bitmap.createBitmap(bitmapForObjectDetection, x, y, width, height)
+        val predictions: MutableList<CharPredictionWithBox> = ArrayList<CharPredictionWithBox>()
+        var begin = 0
+
+        while (begin < width - height) {
```

Comment:
the correlation between `width` and `height` here is really confusing.

Reply:
hah yeah, i added some comments to clear some of this up